File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.java
class Solution {
    public int maxProfit(int[] prices) {
        int[][] dp = new int[prices.length+2][2];
        dp[prices.length][1] = 0;
        dp[prices.length][1] = 0;
        dp[prices.length+1][1] = 0;
        dp[prices.length+1][1] = 0;

        for(int i=prices.length-1; i>=0; i--){
            int profit;
            for(int buy =0; buy<=1; buy++){
                if(buy == 1){
                    profit = Math.max(-prices[i] + dp[i+1][0], dp[i+1][1]);
                }
                else{
                    profit = Math.max(prices[i] + dp[i+2][1],
                                    dp[i+1][0]);
                }
                dp[i][buy] = profit;
            }
        }

        return dp[0][1];
    }
}

File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        n = len(prices)
        self.res = -1

        def choose(idx, haveStock, profit):
            if idx >= n:
                self.res = max(self.res, profit)
                return 

            #buy
            pick = skip = 0
            if haveStock == 0:
                pick = max(choose(idx+1, 1, profit-prices[idx]), choose(idx+1, 0, profit))

            else:
                skip = max(choose(idx+2, 0, profit+prices[idx]), choose(idx+1, 1, profit))

            return max(skip, pick)

        # choose(0, 0, 0)
        # 

        dp = [[-1 for _ in range(n+1)] for _ in range(2)]

        dp[0][n] = 0
        dp[1][n] = 0

        for i in range(n-1, -1, -1):

            #Buy = 0 -> buy or not buy
            dp[0][i] = max(dp[1][i+1]-prices[i], dp[0][i+1])

            #Sell
            dp[1][i] = max(dp[0][min(i+2, n)]+prices[i], dp[1][i+1])   


        return dp[0][0]


        